#' Essential Simulation Parameter Generator
#'
#' @description
#' This function generates an object of type \code{\link{SimParam}} with the essential parameters needed to perform the simulation.
#' After running this function, other methods from \code{\link{SimParam}} can be used to define additional global parameters not included here.
#' Saving this object as \var{SP} will allow it to be accessed by other functions with default parameters.
#'
#' @param founder An object of class \code{\link{MapPop-class}} generated by the function \code{\link{setFounderG}}. If \code{NULL}, the function retrieves the founder haplotypes from the system.
#' @param minQTLchr A value or vector specifying the minimum number of QTL per chromosome. \emph{Default} is 1.
#' @param nSnpChr A value or vector specifying the number of SNPs per chromosome. If \code{NULL}, the \emph{Default} is 100,000 for \dQuote{RABBIT} (based on the 200K Affymetrix Axiom OrcunSNP array \cite{ThermoFisher Scientific}) and 1724 for \dQuote{CATTLE} (based on the Bovine SNP50 v2 BeadChip from \cite{Illumina}). For \dQuote{GENERIC}, a value must be specified.
#' @param minSnpChr A value or vector specifying the number of segregating sites for SNPs per chromosome. If \code{NULL}, it is assumed to be half of \var{nSnpChr}.
#' @param minSnpFreq The minimum allowable frequency for SNP loci. No minimum SNP frequency is applied if \code{NULL}.
#' @param rndSeed A seed to initialize random sampling. If \code{NULL}, a seed is randomly set.
#'
#' @usage
#' essentialSP(founder = NULL,
#'             minQTLchr = 1,
#'             nSnpChr = NULL,
#'             minSnpChr = NULL,
#'             minSnpFreq = NULL,
#'             rndSeed = NULL)
#'
#' @return An object of class \code{\link{SimParam}}.
#' @export
#' @import AlphaSimR
#'
#' @examples
#' \dontrun{
#' # Set genetic parameters
#' SP <- essentialSP(founder = founderG, nSnpChr = 10000, minSnpFreq = 0.05)
#' }
essentialSP <- function(founder = NULL,minQTLchr = 1, nSnpChr = NULL, minSnpChr = NULL, minSnpFreq = NULL,rndSeed=NULL){

  if(is.null(founder)){
    founder <- get("founderG", envir = .GlobalEnv)
  }

  if(is.null(rndSeed)){
    rndSeed <- sample(c(1000:2000),1)
  }

  globalSP <- get("gSP", envir = .GlobalEnv)

  if(is.null(globalSP)){
    stop("globalSP is not found")
  }


  if(is.null(nSnpChr)){
    if(globalSP$animal == "GENERIC"){
      stop("If you are simulating a 'GENERIC' genome, the number of SNP per chromosome is required")
    }else{
      if(globalSP$animal == "CATTLE"){
        nSnpChr <- 1724
      }
      if(globalSP$animal == "RABBIT"){
        nSnpChr <- 10000
      }
    }

  }


  if(is.null(minSnpChr)){
    minSnpChr <- round(nSnpChr*0.5)
  }

  set.seed(rndSeed)
  SP = SimParam$new(founder)
  SP$setSexes("yes_sys")
  set.seed(rndSeed)
  SP$addTraitA(nQtlPerChr = globalSP$nQTLchr,
               mean = 0,
               var = (globalSP$varP*globalSP$h2),
               gamma = T,
               shape = 0.2)

  set.seed(rndSeed)
  SP$restrSegSites(minQtlPerChr = minQTLchr,
                   minSnpPerChr = minSnpChr,
                   overlap = F,
                   minSnpFreq = minSnpFreq)
  set.seed(rndSeed)
  SP$addSnpChip(nSnpPerChr = nSnpChr,minSnpFreq = minSnpFreq)

  return(SP)
}

